services/fxaccounts/RustFxAccount.js

```diff
@@ -390,23 +402,66 @@ class RustFxAccount {
    * @param {[DeviceCapability]} supportedCapabilities
    */
   async ensureCapabilities(supportedCapabilities) {
-    return promisify(this.bridge.ensureCapabilities, supportedCapabilities);
+    return this._promisify(
+      this.bridge.ensureCapabilities,
+      supportedCapabilities
+    );
   }
-}
 
-function promisify(func, ...params) {
-  return new Promise((resolve, reject) => {
-    func(...params, {
-      // This object implicitly implements
-      // `mozIFirefoxAccountsBridgeCallback`.
-      handleSuccess: resolve,
-      handleError(code, message) {
-        let error = new Error(message);
-        error.result = code;
-        reject(error);
-      },
+  _promisify(func, ...params) {
+    const fxa = this;
+    return new Promise((resolve, reject) => {
+      func(...params, {
+        // This object implicitly implements
+        // `mozIFirefoxAccountsBridgeCallback`.
+        handleSuccess(res) {
+          // Get and write the FxA state.
+          // Very much prototype quality code.
+          // There's a high chance the JS event loop will mess up everything,
+          // such as doing OP1 "write-state" AFTER OP2 "write-state".
```

Comment:
Because of this, we might have to guard the `RustFxAccount` ops in a promise queue like we've done on other occasions on Desktop, what do you think @mhammond?

Reply:
Interesting. IIUC we do get queue-like behaviour from `punt`, which ensures that each operation *at the rust level* is executed serially and in order. I wonder if we can also convince it to guarantee calling these callbacks in order rather than having to add another layer of queueing here.

Reply:
Yeah I think it's a good idea, we can access the password manager from C++ I believe.

Reply:
(and even from Rust using XPCOM magic)

Reply:
Sorry I missed this - but yeah, I think the order of these callbacks being *called* can be deterministic, but as soon as the JS code in the callbacks start `await`ing etc, then all bets are off. These notifications aren't async, which we faced in sync too, so we have a "promise queue" there, but that's more about ensuring they are all finished at certain points than about guaranteeing the order of them resolving. A strict ordering queue is tricky because at some points we explicitly chose to let a promise complete in the background - just grep for "background" in services/fxaccounts - often actual persisting of data is done in the background, 

Note sure I'm answering the question, but I'm strongly sympathizing with the problem :)

Reply:
I've moved the persistence to the Rust layer as suggested by @rfk in https://github.com/a-s-dev/gecko-dev/pull/2/commits/81d50947d19a927e50a4cda65aea9c92babcc17c.